corehq/apps/app_manager/suite_xml/post_process/remote_requests.py

```diff
@@ -254,15 +265,74 @@ def build_query_prompts(self):
 
     def build_stack(self):
         stack = Stack()
-        frame = PushFrame()
+        rewind_if = None
+        if module_uses_smart_links(self.module):
+            user_domain_xpath = session_var(COMMCARE_PROJECT, path="user/data")
+            # For case in same domain, do a regular case claim rewind
+            rewind_if = self._get_case_domain_xpath().eq(user_domain_xpath)
+            # For case in another domain, jump to that other domain
+            frame = PushFrame(if_clause=self._get_case_domain_xpath().neq(user_domain_xpath))
+            frame.add_datum(StackJump(
+                url=Text(
+                    xpath=TextXPath(
+                        function=self.get_smart_link_function(),
+                        variables=self.get_smart_link_variables(),
+                    ),
+                ),
+            ))
+            stack.add_frame(frame)
+        frame = PushFrame(if_clause=rewind_if)
         frame.add_rewind(QuerySessionXPath(self.case_session_var).instance())
         stack.add_frame(frame)
         return stack
 
+    def get_smart_link_function(self):
+        app_id = self.app.upstream_app_id if is_linked_app(self.app) else self.app.origin_id
+        url = absolute_reverse("session_endpoint", args=["---", app_id, self.module.session_endpoint_id])
+        prefix, suffix = url.split("---")
+        params = ""
+        argument_ids = self.endpoint_argument_ids
+        if argument_ids:
+            params = f", '?{argument_ids[-1]}=', ${argument_ids[-1]}"
+            for argument_id in argument_ids[:-1]:
+                params += f", '&{argument_id}=', ${argument_id}"
```

Comment:
this is hard to grok but a comment with an example above would help:
```
# URL params need to be evaluated in xpath:
#    concat('?arg1=', $arg1, '&arg2=', $arg2)
#  after xpath eval:
#    '?arg1=va1&arg2=val2'
```

Side note: an xpath function to build URLs that also encoded params would be great

Reply:
260ef14a3c0

Like a function that would build the URL on the CC side? That'd be cool.